### PR TITLE
NaN-aware coord compare

### LINKED ISSUE #2327: NaN value(s) in Dataset coordinates break open_hdf5
If a `Dataset` contains a coordinate with a `numpy.nan` then it can not be retrieved after storage in a HDF5 file.
Such a coordinate may be, e.g., the result of a center-of-mass calculation for a quantity which is all zeros.

```python
import scipp
import numpy
x = scipp.Variable(values=[0, 1, numpy.nan, 3], dims=['x'], unit='m')
y = scipp.Variable(values=[3, 2, numpy.nan, 2], dims=['x'], unit='counts')
i = scipp.Variable(values=[0.1, 0.2, 0.3, 0.4], dims=['x'], unit='amp')

ds = scipp.Dataset(data={'y': y, 'i': i}, coords={'x': x})
ds.to_hdf5('test_scipp_dataset_nan_coord.hdf5')

scipp.io.open_hdf5('test_scipp_dataset_nan_coord.hdf5')
```
throws a `VariableError` with message
```python
VariableError: Expected
<scipp.Variable> (x: 4)    float64              [m]  [0.000000, 1.000000, nan, 3.000000], got
<scipp.Variable> (x: 4)    float64              [m]  [0.000000, 1.000000, nan, 3.000000].
```

A `Dataset` with only one `data` entry (or a `DataArray`) and one or more NaN values in a coordinate can be saved and re-loaded.
The problem is apparently in a comparison between `DataArray` coordinates and may be accessible elsewhere.

I would have guessed that combining `DataArray`s with the same coordinates would raise the same error, but
```python
day = scipp.DataArray(y, coords={'x': x})
dai = scipp.DataArray(i, coords={'x': x})
scipp.Dataset(data={'y': day, 'i': dai})
```
does not.

I think this is a problem with comparing coordinates before combing the constituent `DataArray`s since `numpy.nan == numpy.nan` is always `False`. Would it be possible to change the coordinate-comparison to treat two NaN values as equal?

### PR DESCRIPTION
**Pull request by SimonHeybrock**
Fixes #2327. Changes are:

- Binary operations comparing coords treat NAN values as equal.
- This includes attributes.

The implementation is a bit involved, since we need to handle nested variables, data arrays, and datasets.

I have little confidence that I found *all* cases where coords (or attrs) are compute using `==` (or `!=`) but should actually be changed to `equals_nan`. I suspect we will have to update more operations in the future, as we find issues.

### CODE REVIEW

File: lib/dataset/data_array.cpp [hunk 1]
```diff
@@ -102,6 +102,16 @@ bool operator!=(const DataArray &a, const DataArray &b) {
   return !operator==(a, b);
 }
 
+bool equals_nan(const DataArray &a, const DataArray &b) {
+  if (!equals_nan(a.coords(), b.coords()))
+    return false;
+  if (!equals_nan(a.masks(), b.masks()))
```
Comment (jl-wynen) [inline lib/dataset/data_array.cpp:108]: Do we allow dtypes other than bool in masks? What would NaN mean in a mask?

File: lib/variable/include/scipp/variable/bin_array_model.h [hunk 2]
```diff
@@ -118,6 +120,13 @@ bool BinArrayModel<T>::equals(const Variable &a, const Variable &b) const {
          equals_impl(a.values<bucket<T>>(), b.values<bucket<T>>());
 }
 
+template <class T>
+bool BinArrayModel<T>::equals_nan(const Variable &a, const Variable &b) const {
+  // TODO This implementation is slow since it creates a view for every bucket.
```
Comment (jl-wynen) [inline lib/variable/include/scipp/variable/bin_array_model.h:125]: How slow? Open an issue?

File: lib/variable/include/scipp/variable/element_array_variable.tcc [hunk 3]
```diff
@@ -108,6 +108,18 @@ bool ElementArrayModel<T>::equals(const Variable &a, const Variable &b) const {
           equals_impl(a.variances<T>(), b.variances<T>()));
 }
 
+/// Helper for implementing Variable::operator==.
+///
+/// This method is using virtual dispatch as a trick to obtain T, such that
+/// values<T> and variances<T> can be compared.
```
Comment (jl-wynen) [inline lib/variable/include/scipp/variable/element_array_variable.tcc:114]: Does it? Looks like the caller has to find `T`.

File: lib/variable/include/scipp/variable/element_array_model.h [hunk 4]
```diff
@@ -31,6 +32,29 @@ bool equals_impl(const T1 &view1, const T2 &view2) {
...
+    return std::equal(
+        view1.begin(), view1.end(), view2.begin(), view2.end(),
+        [](const auto &a, const auto &b) { return equals_nan_impl(a, b); });
+  else
+    return std::equal(
+        view1.begin(), view1.end(), view2.begin(), view2.end(),
+        [](const auto &x, const auto &y) {
+          if constexpr (std::is_floating_point_v<std::decay_t<decltype(x)>>) {
+            using numeric::isnan;
+            if (isnan(x) && isnan(y))
```
Comment (jl-wynen) [inline lib/variable/include/scipp/variable/element_array_model.h:51]: Why is this needed? Does `equals_nan(float, float)` not contain this check?

File: lib/dataset/data_array.cpp (hunk 1 above)
Comment (SimonHeybrock) [reply lib/dataset/data_array.cpp:108]: There is no limitation of what can be in a mask, but binary operations (or reductions) raise an exception since you cannot `OR` anything but bool.

File: lib/variable/include/scipp/variable/bin_array_model.h (hunk 2 above)
Comment (SimonHeybrock) [reply lib/variable/include/scipp/variable/bin_array_model.h:125]: This is not a new comment (see a couple lines above). It has not showed up in practice (maybe because we don't have bins in coords, and don't compare things much other wise outside unit tests?).

File: lib/variable/include/scipp/variable/element_array_model.h (hunk 4 above)
Comment (SimonHeybrock) [reply lib/variable/include/scipp/variable/element_array_model.h:51]: It just uses `operator==`, but you are right that it may be clearer to move this bit into that helper function.

File: lib/variable/include/scipp/variable/bin_array_model.h (hunk 2 above)
Comment (jl-wynen) [reply lib/variable/include/scipp/variable/bin_array_model.h:125]: I just wanted to make sure this isn't lost in the source code but properly traced if it were important.

File: lib/variable/include/scipp/variable/element_array_variable.tcc (hunk 3 above)
Comment (jl-wynen) [reply lib/variable/include/scipp/variable/element_array_variable.tcc:114]: That happens before the function is even called.
A bit nit picky and I am therefore resolving this conversation.

**Review by jl-wynen — APPROVED** [review]

**Comment by jl-wynen** [comment]
I cannot think of any cases you might have missed on the top of my head. But this PR should improve the existing situation even if it missed some cases.
